src/allmydata/storage/http_server.py

```diff
@@ -154,3 +164,116 @@ def _cbor(self, request, data):
+        info = loads(request.content.read())
+        upload_key = authorization[Secrets.UPLOAD]
+
+        if storage_index in self._uploads:
+            # Pre-existing upload.
+            in_progress = self._uploads[storage_index]
+            if in_progress.upload_key == upload_key:
+                # Same session.
+                # TODO add BucketWriters only for new shares
+                pass
+            else:
+                # New session.
+                # TODO cancel all existing BucketWriters, then do
+                # self._storage_server.allocate_buckets() with given inputs.
+                pass
+        else:
+            # New upload.
+            already_got, sharenum_to_bucket = self._storage_server.allocate_buckets(
+                storage_index,
+                renew_secret=authorization[Secrets.LEASE_RENEW],
+                cancel_secret=authorization[Secrets.LEASE_CANCEL],
+                sharenums=info["share-numbers"],
+                allocated_size=info["allocated-size"],
+            )
+            self._uploads[storage_index] = StorageIndexUploads(
+                shares=sharenum_to_bucket, upload_key=authorization[Secrets.UPLOAD]
+            )
+            return self._cbor(
+                request,
+                {
+                    "already-have": set(already_got),
+                    "allocated": set(sharenum_to_bucket),
+                },
+            )
+
+    @_authorized_route(
+        _app,
+        {Secrets.UPLOAD},
+        "/v1/immutable/<string:storage_index>/<int:share_number>",
+        methods=["PATCH"],
+    )
+    def write_share_data(self, request, authorization, storage_index, share_number):
+        """Write data to an in-progress immutable upload."""
+        storage_index = si_a2b(storage_index.encode("ascii"))
+        content_range = request.getHeader("content-range")
+        if content_range is None:
+            offset = 0
+        else:
+            offset = int(content_range.split()[1].split("-")[0])
+
+        # TODO basic checks on validity of start, offset, and content-range in general. also of share_number.
+        # TODO basic check that body isn't infinite. require content-length? or maybe we should require content-range (it's optional now)? if so, needs to be rflected in protocol spec.
+
+        data = request.content.read()
+        try:
+            bucket = self._uploads[storage_index].shares[share_number]
+        except (KeyError, IndexError):
+            # TODO return 404
+            raise
+
+        finished = bucket.write(offset, data)
+
+        # TODO if raises ConflictingWriteError, return HTTP CONFLICT code.
+
+        if finished:
+            bucket.close()
+            request.setResponseCode(http.CREATED)
+        else:
+            request.setResponseCode(http.OK)
+
+        # TODO spec says we should return missing ranges. but client doesn't
+        # actually use them? So is it actually useful?
```

Comment:
Ah, yes.

Reply:
After further thought:

1. The client needs to keep track of what it's sent anyway, just to make sure it's all uploaded (it shouldn't rely on this, because that introduces a bunch more latency).
2. If the client crashes, it needs to track some persistent state anyway (upload secret).
3. The failure mode for not having this is... uploading a bit extra. But the code gets simpler on both client and server.

So I propose just not doing this, and if it ever turns out to be, useful we can add an extra endpoint which is "show status of in-progress upload".

Reply:
(1) makes sense - albeit, for in-memory tracking.  The client probably wants to start uploading the 2nd chunk before the 1st finishes (maybe it even keeps N of them running in parallel all the time).  However, I don't think that (2) follows.  I'm still pretty sure that the upload secret is going to be a deterministic function of other local state the client holds.  Do we just need to design that to get rid of the uncertainty about it?  So, given this information in the response, I don't believe the client will actually need persistent state to be able to resume an upload.  Beyond that, even if the client keeps local state, it has no way of knowing whether the server has decided to give up on the transfer and throw away the data or not.  If it uses only local persistent state then it will "finish" the upload while the beginning is now missing.  It can probably infer that this is the case when the last write doesn't return a CREATED code but that's extra client side logic too.

For (3) we can't say if it is a little bit or not.  If I have a 20GB share to upload and I get halfway through, it would be pretty sad for my client to have to start over again when we could pretty easily avoid it.  To take this to the extreme, maybe my setup is flaky enough that I can *never* complete an upload in one try.  Persistent local state is another solution to this but between "the function returns enough information to finish the process" and "keep local state" my personal preference is definitely for the former because it introduces fewer difficult requirements on the client.

Does the server get that much more complicated by having to return this information?  Doesn't it already have to keep track of it to make sure that no single location in the sharefile can have different bytes written to it?

Reply:
It's not that big a deal, I'll add it. The only potential problem is https://github.com/mlenzen/collections-extended/issues/172 and looks like that'll be fixed in near future.